teslajsonpy/controller.py

```diff
@@ -948,78 +1002,163 @@ async def _get_and_process_energysite_data(energysite_id: Text) -> None:
 
             return any(await asyncio.gather(*tasks))
 
-    def get_climate_params(self, car_id):
+    def get_climate_params(self, car_id: Text = None, vin: Text = None) -> Dict:
         """Return cached copy of climate_params for car_id."""
-        vin = self._id_to_vin(car_id)
-        if vin:
+        if car_id and not vin:
+            vin = self._id_to_vin(car_id)
+        if vin and vin in self.__climate:
             return self.__climate[vin]
-        return {}
+        return self.__climate
 
-    def set_climate_params(self, car_id: Text, params: Dict) -> None:
+    def set_climate_params(
+        self, car_id: Text = None, vin: Text = None, params: Dict = None
+    ) -> None:
         """Set climate_params for car_id."""
-        vin = self._id_to_vin(car_id)
+        params = params or {}
+        if car_id and not vin:
+            vin = self._id_to_vin(car_id)
         if vin:
             self.__climate[vin] = params
 
-    def get_charging_params(self, car_id):
+    def is_climate_on(self, car_id: Text = None, vin: Text = None) -> bool:
+        """Return true if climate is on."""
+        if car_id and not vin:
+            vin = self._id_to_vin(car_id)
+        if vin and vin in self.__climate:
+            return self.get_climate_params(vin=vin).get("is_climate_on")
+        return False
+
+    def get_charging_params(self, car_id: Text = None, vin: Text = None) -> Dict:
         """Return cached copy of charging_params for car_id."""
-        vin = self._id_to_vin(car_id)
-        if vin:
+        if car_id and not vin:
+            vin = self._id_to_vin(car_id)
+        if vin and vin in self.__charging:
             return self.__charging[vin]
-        return {}
+        return self.__charging
 
-    def set_charging_params(self, car_id: Text, params: Dict) -> None:
+    def set_charging_params(
+        self, car_id: Text = None, vin: Text = None, params: Dict = None
+    ) -> None:
         """Set charging_params for car_id."""
-        vin = self._id_to_vin(car_id)
+        params = params or {}
+        if car_id and not vin:
+            vin = self._id_to_vin(car_id)
         if vin:
             self.__charging[vin] = params
 
-    def get_power_params(self, site_id):
+    def charging_state(self, car_id: Text = None, vin: Text = None) -> Text:
+        """Return charging state for a single vehicle."""
+        if car_id and not vin:
+            vin = self._id_to_vin(car_id)
+        if vin and vin in self.__charging:
+            return self.get_charging_params(vin=vin).get("charging_state")
+        return None
```

Comment:
Same question about this return. Should it be None or empty dict. Regardless of what we choose, it should be consistent so we don't confuse ourselves later.

Reply:
In this function, I think `None` is the correct return value.  

The functions `is_car_online`, `is_climate_on`, `shift_state`, `charging_state` etc. are shortcuts for getting a single value (or boolean) for one specific car.

I have tried to be consistent, in the sense that the `get_`-functions returns parameters for a single car if vin or car_id is provided and a full dict for all cars if none of these are given.

The `is_` functions are "questions" that can be answered `true` or `false`.

And then the `_state` functions (`shift_state`, and `charging_state`) returns a string with the actual state of a single parameter (given a vin or car_id) or `None`.

(But this made me have another look at the code, and I found that the function "in_gear" should really be "is_in_gear", so I just made another commit to fix that).

I will also update the docstrings to be more clear.